### PR TITLE
Improved German translation

### PR DESCRIPTION
**Pull request by H0mebrewL0ver**
I have considered the feedback in PR https://github.com/V10lator/NUSspli/pull/139. I assume you understand German, so I will go into more detail.

- The imperative form should only be used in German, if the user has to do something (action required)
Example: Press A to continue = Drücke A zum Fortfahren
- If the string is used as a button or the purpose of the string is to inform the user about something, the imperative form should never be used in German
Example: Downloading = Wird heruntergeladen … (not "Downloade" or "Lade herunter")
- That is also the reason why I have added some ellipsis. The ellipsis indicate a process
- Furthermore, [there has to be a space between the ellipsis and the word in German](https://www.duden.de/sprachwissen/sprachratgeber/Auslassungspunkte). I have used a no-break-space (geschütztes Leerzeichen) for this. Of course we could use a "normal" space if this causes any problems
- Words after "zum" are usually capitalized ([Substantivierung](https://www.duden.de/sprachwissen/rechtschreibregeln/Gro%C3%9F-%20und%20Kleinschreibung#D82)).
- [There has to be a hyphen for some words](https://www.duden.de/sprachwissen/rechtschreibregeln/bindestrich#D26) like ZIP-Datei or Fake-Datei. Please have a look at line 4 (Fake-<title.tik>-Datei), grammar rules for German demand hyphen's here, but maybe there is a better solution. I also noticed that this is the only string with <title.tik>. In other strings it is just title.tik, is there a reason for this?

I have noticed that some strings are not included in the localization file. "Online updates", "Operation:" and "Logs" are missing. I have added them with the format of other strings. Please clarify if this works.

There are also some strings that seem to use placeholders (line 125) and I am not sure I have accounted for everything. I would also like to get some context for:
- line 25 "Searching on"
- line 114-130 are there placeholders used?
- line 144 what is MCP?
- line 186 "unknown"

Please leave feedback for any other things you have noticed!

### CODE REVIEW

**Comment by V10lator** [comment]
Thanks for this. I like some of the changes while others I dislike. For example you might be right with the hypen between some words but I guarantee you every german citizen will understand what `Online updates` means, so it has not been added to the translation to save space and make it more simple to maintain.
Same goes for "Logs" and here I even like it more than "Protokoll" cause who the f** says for example "Serverprotokoll" in german IT these days, for example?

> There are also some strings that seem to use placeholders (line 125)

Line 125 of what file? At https://github.com/V10lator/NUSspli/blob/master/data/locale/German.json it's an empty line (should help getting an overview, not a placeholder). At https://github.com/V10lator/NUSspli/blob/master/meta/baselocale.json it's a line which is indeed translated in the german translation, so also not a placeholder.

> line 25

Empty line in both files:
https://github.com/V10lator/NUSspli/blob/master/data/locale/German.json
https://github.com/V10lator/NUSspli/blob/master/meta/baselocale.json
Are you talking about line 24? If so what exactly do you want to know? This string is used here, for example:
grafik
And there's also something weird in your translation: All looks good on that screen except `Drücke (X) zur SD-Karte zu wechseln` ... Not sure if there's still enough space for a `um`. Also it's called `SD` everywhere, you changed just this single entry to `SD-Karte`.

Rest with line numbers I won't reply to until clarification about what file we're even talking.

> what is MCP?

A very big program doing a lot of administrative tasks in the operation system(s) (yes, it runs partly on CafeOS and partly on IOUS) Nintendo created for the Wii U. Seeing what it does I guess Nintendo devs watched too much Tron and this stands indeed for `Master Control Program`. NUSspli utilises MCP for various things, including the title installations itself.

Do not start to change anything on this PR for now, please. I will give more replies later, so wait for them. please. :)

**Comment by V10lator** [comment]
BTW: You can find test builds including your translation here: https://github.com/V10lator/NUSspli/actions/runs/10006277161

**Comment by V10lator** [comment]
One last thing then I'll go into individual points of your changes (which needs more time) :
This `Protokoll` vs. `Logs` thing: We should just accept that languages change and that german language incooperates words like these. One example is "Hast du die Logs schon durchgesehen?" in german IT. Same goes btw the other round: "What kind of kindergarden is this" is a valid english sentence, even if 'Kinder' is a german word not used in english anywhere else (and 'Kind' even has a completey different meaning in engish and in german).

File: data/locale/German.json [hunk 1]
```diff
@@ -1,186 +1,189 @@
...
-    "drive?": "Laufwerk löschen?",
-
-    "Uninstalling": "Deinstalliere",
-    "Preparing...": "Vorbereitung...",
+    "hours": "Stunden",
+    "minutes": "Minuten",
+    "seconds": "Sekunden",
+
+    "Downloading": "Wird heruntergeladen ...",
+    "Preparing": "Vorbereitung läuft ...",
```
Comment (V10lator) [inline data/locale/German.json:115]: This somehow looks weird, too (see screenshot above).

File: data/locale/German.json [hunk 2]
```diff
@@ -1,186 +1,189 @@
...
+    "Preparing the download of": "Vorbereitung fürs Herunterladen von",
+    "Creating fake title.tik": "Fake-title.tik-Datei wird erstellt ...",
+    "Creating CERT": "CERT wird erstellt ...",
+
+    "Installing": "Wird installiert ...",
```
Comment (V10lator) [inline data/locale/German.json:120]: Didn't look at it on.-screen yet but I bet this will look weird, too.

File: data/locale/German.json [hunk 3]
```diff
@@ -1,186 +1,189 @@
...
+    "Preparing. This might take some time. Please be patient.": "Vorbereitung läuft. Dies kann einen Moment dauern ...",
+    "Cancelling installation.": "Installation wird abgebrochen ...",
+    "Please wait...": "Bitte warte ...",
+
+    "Do you really want to uninstall": "Möchtest du",
+    "from your": "wirklich von deinem",
+    "drive?": "Laufwerk deinstallieren?",
+
+    "Uninstalling": "Wird deinstalliert ...",
```
Comment (V10lator) [inline data/locale/German.json:129]: This might look weird, too.

File: data/locale/German.json [hunk 4]
```diff
@@ -1,186 +1,189 @@
...
+    "Preparing...": "Vorbereitung läuft ...",
     "Downloaded successfully!": "Erfolgreich heruntergeladen!",
     "Installed successfully!": "Erfolgreich installiert!",
-    "Uninstalled successfully!": "Deinstallation erfolgreich!",
+    "Uninstalled successfully!": "Erfolgreich deinstalliert!",
```
Comment (V10lator) [inline data/locale/German.json:133]: Why this change?

File: data/locale/German.json [hunk 5]
```diff
@@ -1,186 +1,189 @@
...
     "Queue finished successfully!": "Warteschlange erfolgreich abgearbeitet!",
 
-    "Could not open path": "Konnte Pfad nicht öffnen:",
-    "Invalid title.tmd file!": "Fehlerhafte title.tmd Datei!",
-    "Couldn't save config file!\nYour SD card might be write locked.": "Konnte Konfigurationsdatei nicht lesen!\nSD Karte write locked?",
-    "Downloading to NAND is dangerous,\nit could brick your Wii U!\n\nAre you sure you want to do this?": "Download zum NAND ist gefährlich,\nes könnte deine Wii U bricken!\n\nBist du sicher?",
-    "Not enough free space on": "Nicht genug freier Speicher bei",
-    "No title.tmd found at": "Keine title.tmd Datei bei",
+    "Could not open path": "Pfad konnte nicht geöffnet werden",
```
Comment (V10lator) [inline data/locale/German.json:136]: There will be a path following. So are you sure this will be better?

File: data/locale/German.json [hunk 6]
```diff
@@ -1,186 +1,189 @@
...
+    "Files might be corrupt or bad storage medium.\nTry redownloading files or reformat/replace target device": "Möglicherweise sind die Dateien oder das Speichermedium beschädigt.\nVersuch, die Dateien erneut herunterzuladen oder formatiere/ersetze den Datenträger",
+    "Possible USB failure. Check your drives power source.": "Möglicherweise ein USB-Fehler. Überprüf die Stromversorgung deines Laufwerks.",
+    "Files might be corrupt": "Dateien könnten beschädigt sein",
     "Unknown Error": "Unbekannter Fehler",
 
     "Network error": "Netzwerkfehler",
-    "check the network settings and try again": "prüfe die Netzwerkeinstellungen und versuche es nochmal",
-    "SSL error": "SSL Fehler",
-    "check your Wii Us date and time settings": "Prüfe die Datums- und Zeiteinstellungen deiner Wii U",
+    "check the network settings and try again": "Überprüf die Netzwerkeinstellungen und versuch es erneut",
```
Comment (V10lator) [inline data/locale/German.json:161]: `versuch` is not a german word. It's either `versuche` (von `versuchen` or `der Versuch`, see:
https://www.duden.de/rechtschreibung/versuchen
https://www.duden.de/rechtschreibung/Versuch
Now granted the Duden doesn't explicy state `versuche` and I can't remember the grammar rule behind it but just google for `versuche es erneut` and you'll see that it's used everywhere.

`überprüf` also is not a german word. It's `überprüfe`. Same goes for the `Prüf` one line above. granted, you could do both as `Überprüf'` / `Prüf'` but that just looks completely unprofessional.

File: data/locale/German.json [hunk 7]
```diff
@@ -1,186 +1,189 @@
...
-    "The server didn't return any data": "Der Server hat keinerlei Daten zurück geschickt",
-    "I/O error": "I/O Fehler",
+    "Couldn't resolve hostname": "Hostname konnte nicht aufgelöst werden",
+    "Couldn't connect to server": "Verbindung zum Server konnte nicht hergestellt werden",
+    "Operation timed out": "Zeitüberschreitung",
+    "The server didn't return any data": "Der Server hat keine Daten zurückgegeben",
+    "I/O error": "I/O-Fehler",
     "Verification failed": "Verifikation fehlgeschlagen",
-    "Handshake failed": "Handschlag fehlgeschlagen",
+    "Handshake failed": "Handshake fehlgeschlagen",
```
Comment (V10lator) [inline data/locale/German.json:171]: Now I'm very interested in the reason you changed other english words into german but here you did the opposite. Not that I think this is wrong, just curious. ;)

File: data/locale/German.json [hunk 8]
```diff
@@ -1,186 +1,189 @@
...
-    "Press \uE003 to retry": "Drücke \uE003 um es nochmal zu versuchen",
-    "This is a reliable way to brick your console!\nAre you sure you want to do that?": "Dies ist ein sicherer Weg deine Konsole zu beicken!\nBist du sicher das du dies möchtest?",
-    "Are you really sure you want to brick your Wii U?": "Bist du sicher das du deine Konsole zerstören willst?",
-    "You're on your own doing this,\ndo you understand the consequences?": "Du bist selbst verantwortlich wenn du fortfährst,\nverstehst du die Konsequenzen?",
-
-    "Search": "Suche",
-    "unknown": "unbekannten",
+    "Unknown libcurl error": "Unbekannter libcurl-Fehler",
+    "Do you really want to cancel?": "Möchtest du wirklich abbrechen?",
+    "The download of title.tmd failed with error: 404": "Das Herunterladen von title.tmd schlug mit dem Code 404 fehl",
```
Comment (V10lator) [inline data/locale/German.json:175]: The `der title.tmd` was short for `der title.tmd Datei`, so `von title.tmd` sounds wrong to me. Care to elaborate on this change?

File: data/locale/German.json [hunk 9]
```diff
@@ -1,186 +1,189 @@
...
+    "The download returned a result different to 200 (OK)": "Beim Herunterladen wurde eine andere Antwort als 200 (OK) zurückgegeben",
+    "Request failed. Try again": "Anfrage fehlgeschlagen. Versuch es erneut.",
+
+    "Press \uE003 to retry": "Drücke \uE003 zum Wiederholen",
+    "This is a reliable way to brick your console!\nAre you sure you want to do that?": "Du bist auf dem besten Weg, deine Konsole zu zerstören!\nMöchtest du das Risiko eingehen?",
+    "Are you really sure you want to brick your Wii U?": "Möchtest du deine Wii U wirklich zerstören?",
+    "You're on your own doing this,\ndo you understand the consequences?": "Du bist selbst dafür verantwortlich, wenn du fortfährst.\nIst dir das Ausmaß wirklich bewusst?",
+
+    "Search": "Suchen",
+    "unknown": "Unbekannt",
```
Comment (V10lator) [inline data/locale/German.json:186]: image1

**Review by V10lator — COMMENTED** [review]
It might be a real good idea to split this big PR into multiple smaller ones

**Comment by eku** [comment]
May I suggest shortening the translations if necessary, if they are longer, to maintain the arrangement of the text elements.

**Comment by eku** [comment]
zu bekommen -> erhalten ?

**Comment by V10lator** [comment]
> it could still be a problem for other languages.

Each translator decides for himself what to translate and what not. NUSspli doesn't have a rule here but the translators should do whatever fits best to the language.

> This menu option is displayed at the start of the app, but only experienced users would benefit from it.

This is true no matter how we call the option as the logs are for experienced users / debugging purposes only. ;)

> 111-127 (especially line 122 "Willst du wirklich [I guess something will be inserted here] von deinem Laufwerk löschen?")

You can always search the sourcecodes for the english strings to see where and how they are used. For line 111 this would be:
```
$ grep -r "Downloading" src/
src/downloader.c:                strcpy(toScreen, localise("Downloading"));
src/downloader.c:    debugPrintf("Downloading title... tID: %s, tVer: %s, name: %s, folder: %s", tid, titleVer, titleEntry->name, folderName);
src/menu/predownloadMenu.c:    debugPrintf("Downloading TMD...");
src/menu/predownloadMenu.c:                    "Downloading to NAND is dangerous,\n"
```
So let's look into `src/downloader.c` and search for that: https://github.com/V10lator/NUSspli/blob/master/src/downloader.c#L621-L624
What this does is indeed building a string with the localised text, then a space and then a variable (the file name in this case), which you could call using a placeholder. Finally it writes this created string to the screen.

Will do the same for the string at line 122 now, then leave the rest up to you:
```
$ grep -r "Do you really want to uninstall" src/
src/menu/insttitlebrowserMenu.c:        strcpy(toFrame, localise("Do you really want to uninstall"));
```
https://github.com/V10lator/NUSspli/blob/master/src/menu/insttitlebrowserMenu.c#L458-L473
Now we see this builds a larger string, using a lot more things... It creates something like this:
```
Do you really want to uninstall
[NAME OF THE TITLE OR TITLE ID AS FALLBACK]
from your [PLACEHOLDER] drive?

(A) Yes || (B) No
```
Using line 122, line 123, maybe line 47, line 124, line 184 and line 185 from https://github.com/V10lator/NUSspli/blob/master/data/locale/German.json

> I guess, it should be vom MCP instead of von MCP

While being grammatically correct on first glance this somehow sounds unintuitive / weird to me. Will need to think some more about that I guess.

> I have no idea how to do that.

You do that by creating new branches, then doing just some of the work in each branch, then doing PRs for each new branch you created (to the master branch). It's not strictly needed but makes reviewing so much simpler + allows us to already merge some things while still talking about others.

> Could you please show the whole sentence? I don't know why zu bekommen is relevant.

You know how to do this now but one more time:
```
$ grep -r "from MCP" src/
src/installer.c:                sprintf(toScreen, "%s \"%s\" %s: %#010x", localise("Error getting info for"), path, localise("from MCP"), data.err);
```
We don't even need to open the file this time as we already see what#s going on, even if pretty crypted for a none-programmer. The `"%s \"%s\" %s: %#010x"` says write `[PLACEHOLDER STRING] "[PLACEHOLDER STRING]" [PLACEHOLDER STRING]: [PLACEHOLDER HEXADECIMAL VALUE WITH 10 DIGITS, PREFIX WITH 0s IF NECESSARY]`. After that it follows what to use as placeholders: `localise("Error getting info for"), path, localise("from MCP"), data.err`. So the final string this creates will look something like this: `Error getting info for /some/path from MCP: [ERROR CODE IN HEX]`.

> [because Nintendo](https://www.nintendo.com/de-de/Wii-U/Inhalte-zum-Herunterladen/Spiele-erwerben/Spiele-erwerben-672048.html)

I really don't think we have to follow Nintendos wording. I mean we're not affiliated with them and what other homebrew does this? Want to hear the opinion of the other translators on this through as this would be something they have to follow, too, if we decide to go that route.

File: data/locale/German.json (hunk 7 above)
Comment (V10lator) [reply data/locale/German.json:170]: The proper translation really is "Handschlag" (example: https://de.wikipedia.org/wiki/Drei-Wege-Handschlag ). Anyway, as said what you did was not wrong, was just curious.
